File: events/ccppbrasil_13o_encontro/code/sample-1/req.c
//Sample provided by Fabio Galuppo  
//April 2017

//compile:
//clang -I../include -L/usr/local/Cellar/zeromq/4.2.2/lib -lzmq req.c -o ./bin/req.exe

//run:
//./bin/req.exe

#include <zmq.h>
#include <stdio.h>
#include <string.h>
#include "common.h"

int main(int argc, char* argv[]) {
    char zmq_version_string[64];
    get_zmq_version_string(zmq_version_string);
    printf("ZMQ version = %s\n", zmq_version_string);

    void* zmq_context = zmq_ctx_new();
    void* request_socket = zmq_socket(zmq_context, ZMQ_REQ);
    zmq_connect(request_socket, "tcp://localhost:60000");
    
    char* msg = "Hello, World!";
    if (argc > 1)
        msg = argv[1];
    
    zmq_send(request_socket, msg, strlen(msg), 0);

    char buffer[64];
    zmq_recv(request_socket, buffer, sizeof(buffer), 0);
    printf("%s\n", buffer);

    zmq_close(request_socket);
    zmq_ctx_term(zmq_context);
    return 0;
}

File: events/visual_studio_summit_2018/src/sample1/Req/Req.cs
//Sample provided by Fabio Galuppo  
//May 2018

using System;

class Req 
{
    static void Main(String[] args) 
    {
        Console.WriteLine("ZMQ version = {0}", Utils.GetZmqVersion());

        var ctx = ZeroMQ.zmq_ctx_new();
        if (ctx != null)
        {
            var requestSocket = ZeroMQ.zmq_socket(ctx, ZeroMQ.ZMQ_REQ);
            if (requestSocket != null)
            {
                if (-1 != ZeroMQ.zmq_connect(requestSocket, "tcp://localhost:60000")) 
                {
                    String msg = 0 == args.Length ? 
                        "Hello World" : 
                        args[0].Substring(0, Math.Min(args[0].Length, 64));
                    using(var buffer = PinnedBuffer.ASCII(msg)) 
                    {
                        if (-1 != ZeroMQ.zmq_send(requestSocket, buffer, buffer.Length, 0)) 
                        {
                            Console.WriteLine(msg + " sent");
                            if (-1 != ZeroMQ.zmq_recv(requestSocket, buffer, buffer.Length, 0))
                            {
                                Console.WriteLine(PinnedBuffer.ASCII(buffer) + " received back");
                            }
                            else
                            {
                                Console.WriteLine("receive back failed");
                            }
                        }
                        else
                        {
                            Console.WriteLine("send failed");
                        }
                    }
                }
                else
                {
                    Console.WriteLine("connect failed");
                }

                ZeroMQ.zmq_close(requestSocket);
            }
            else
            {
                Console.WriteLine("socket failed");
            }

            ZeroMQ.zmq_ctx_term(ctx);
        }
        else
        {
            Console.WriteLine("context failed");
        } 
    }
}
